rfcs/0017-intensional-store.md

```diff
@@ -0,0 +1,391 @@
+1. The daemon verifies the `$cas`. If it doesn't match, it removes `$cas` and `$cas.m`. Note that a missing or altered `$cas.m` file won't pass validation.
+1. If `$cas.m` exists, it checks that all dependencies are already present in the Store. If not, the path is held for a while and deleted if the dependencies don't appear in time (configurable).
+1. It atomically moves `$cas.m` into `/var/lib/nix`.
+1. It atomically moves `$cas` into `/var/lib/nix`.
+
+Note that to ensure atomicity, `.prepare` and `.stage` need to be on the same filesystem, and either `.stage` or `.daemon` need so be on the same filesystem as the Store.
+
+#### without Store Daemon
+
+Any user with write access to `/var/lib/nix/.stage` and `/var/lib/nix` can install entries. To do so:
+
+1. They prepare entries to be stored in `/var/lib/nix/.stage`, renaming each entry to its `$cas` and `$cas.m`.
+1. They atomically move prepared entries to `/var/lib/nix`, in reverse dependency order, meaning dependencies of an entry are moved first, and `$cas.m` is moved before `$cas`
+
+Note that to ensure atomicity, `.stage` needs to be on the same filesystem as the Store.
+
+Note that when two writers are trying to install the same `$cas` or `$cas.m`, one of them might get an error, but the end result will be the same (as long as the `$cas` is self-valid). So multiple writers can also be on separate hosts, in a trusted setting.
+
+### Verification
+
+A path in the Store is verified by calculating its `$cas` (see above). If the `$cas` doesn't match, the path is moved to `/var/lib/nix/.quarantaine`, where a sysadmin has to investigate.
+
+Any process with write access to `/var/lib/nix` and `/var/lib/nix/.quarantaine` can do this, for example the Store daemon.
+
+### Garbage collection
+
+Garbage collection needs to identify store paths that are not used by anything on any of the systems sharing the same store. Here we propose a simple mechanism for coordination, but any mechanism is acceptable.
+
+- A host with store write access decides to run garbage collection.
+- It checks that `.gc/running_gc` does not exist or contains a very old timestamp, and writes a unique number to `.gc/will_gc`.
+- After waiting long enough to prevent collisions (for example 10 seconds), it reads `.gc/will_gc` and verifies it contains the unique number it wrote.
+- It clears out `.gc/` except for the file `.gc/will_gc` and adds the file `.gc/running_gc` containing the current timestamp.
+- While it waits for other hosts, it checks the Store for `$cas.m` files that don't have a matching `$cas`.
+- Each host's store daemon monitors `.gc/running_gc` at some interval, for example 1 minute.
+- While this file exists, the daemon must record its root `$cas` entries, by creating 0-length files named `.gc/$cas`.
+- The writer waits long enough for all the hosts to record their GC roots, for example 10 minutes.
```

Comment:
Not sure how good an idea once a minute pull polling is.

Not sure how all this should behave in case of an hour-long net split, if the design is for distributed use.

Not sure 10 minutes are always enough time to list all the roots even without net split. Should there be support for something like positive reply and request for an extension?

I would probably assign each host a sub-namespace and let the host keep track of GC roots there, somehow. Disappeared hosts can be wiped as an administrative action. Maybe make each host update something once in an hour, if it is reachable, so one can easily query for hosts not seen for a day or a week.

Reply:
Ah yes, that's better. Each host claims a subdir under `.gc/` to use for root marking, and any directory that's not touched in (for example) a week is removed.

To prevent removing paths that were just installed, each host should touch their dir before installing, and GC is not done if any of the directories are too recent (for example an hour).

Multiple GC writers would perform the same actions, so further coordination isn't necessary.

I like :+1:

Reply:
I wonder how to handle the profile pruning options on nix-collect-garbage. Add some policy file to `.gc` that daemons must adhere to for all their profiles, and let them manage the profiles?

Daemons won't necessarily have access to user profiles, but they could refuse to add out-of-policy profile entries to the GC roots.

Reply:
In principle, profile pruning is supposed to be triggered on the machine the user uses, absent an administrative override? And then the root list is updated in the machine's GC-roots directory.

Reply:
Right - but there is a loss of semantics at the level of the GC roots. We lose the information that a given `$cas` is 7 genererations behind the latest generation and 5 months old. It could be retained by making a hierarchical tree of symlinks, complicating matters a bit.

OTOH, a malicious user could just generate 1000s of different profiles and auto-roots, and then maybe the local daemon should implement some limits, or just don't allow this user to install store entries.

So maybe for now just assume that people will be responsible profile pruners.

Reply:
A malicious user could generate just one profile that is bad enough both in storage and in inodes…

Quotas for GC roots for each remote system, and for each user inside system, are the easy part anyway. Accounting for space usage is trickier…

Reply:
This GC could probably be it's own RFC. I think the proposed approach is troubling, I think the proactive tracking that @7c6f434c proposed is much closer to the right approach. However I'd have to read up on POSIX to see what ordering guarantees we can rely on.